{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio.SeqIO import QualityIO\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import gzip\n",
    "import glob\n",
    "import re\n",
    "from DMS_utils import dna_rev_comp, translate_dna2aa\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import pickle as pkl\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy import stats\n",
    "import os.path\n",
    "from matplotlib.lines import Line2D\n",
    "import json\n",
    "import shutil\n",
    "from functions_ import *\n",
    "from plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CGCCGCATGGAAGCGATTAACGAAAGCAGCGGTTTAGCCACAACGCTGGAACGCATTGAAAAGAATTTCGTAATCACAGACCCGCGCCTTCCCGACAATCCAATTATTTTTGCGTCCGATAGCTTCCTGCAATTAACCGAATACAGCCGCGAAGAAATTCTGGGTCGTAATTGTCGCTTCCTTCAGGGGCCAGAGACTGACCGTGCTACGGTACGCAAAATCCGCGACGCAATCGACAATCAAACGGAAGTCACGGTTCAGTTGATTAACTATACGAAGAGCGGAAAAAAATTCTGGAATTTATTTCACTTGCAGCCTATGCGTGACCAGAAGGGCGATGTCCAGTATTTCATTGGCGTTCAGCTTGATGGTACCGAGCATGTTCGCGATGCTGCGGAGCGTGAAGGTGTAATGTTAATTAAAAAGACTGCTGAAAACATTGATGAGGCGGCCAAAGGGAGCCTGCATCCGCCGATGGATAACCGCGTGCG'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"cgccgcatggaagcgattaacgaaagcagcggtttagccacaacgctggaacgcattgaaaagaatttcgtaatcacagacccgcgccttcccgacaatccaattatttttgcgtccgatagcttcctgcaattaaccgaatacagccgcgaagaaattctgggtcgtaattgtcgcttccttcaggggccagagactgaccgtgctacggtacgcaaaatccgcgacgcaatcgacaatcaaacggaagtcacggttcagttgattaactatacgaagagcggaaaaaaattctggaatttatttcacttgcagcctatgcgtgaccagaagggcgatgtccagtatttcattggcgttcagcttgatggtaccgagcatgttcgcgatgctgcggagcgtgaaggtgtaatgttaattaaaaagactgctgaaaacattgatgaggcggccaaagggagcctgcatccgccgatggataaccgcgtgcg\".upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "## reload the evaluation functions\n",
    "from importlib import reload \n",
    "import plotting\n",
    "reload(plotting)\n",
    "from plotting import *\n",
    "import functions_\n",
    "reload(functions_)\n",
    "from functions_ import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "### define the necessary variables\n",
    "\n",
    "base_dir = os.getcwd() \n",
    "\n",
    "LOV_gene = \"TTAGCCACAACGCTGGAACGCATTGAAAAGAATTTCGTAATCACAGACCCGCGCCTTCCCGACAATCCAATTATTTTTGCGTCCGATAGCTTCCTGCAATTAACCGAATACAGCCGCGAAGAAATTCTGGGTCGTAATTGTCGCTTCCTTCAGGGGCCAGAGACTGACCGTGCTACGGTACGCAAAATCCGCGACGCAATCGACAATCAAACGGAAGTCACGGTTCAGTTGATTAACTATACGAAGAGCGGAAAAAAATTCTGGAATTTATTTCACTTGCAGCCTATGCGTGACCAGAAGGGCGATGTCCAGTATTTCATTGGCGTTCAGCTTGATGGTACCGAGCATGTTCGCGATGCTGCGGAGCGTGAAGGTGTAATGTTAATTAAAAAGACTGCTGAAAACATTGATGAGGCGGCCAAA\"\n",
    "\n",
    "LOV_gene_corrected = 'TTAGCCACAACGCTGGAACGCATTGAAAAGAATTTCGTAATCACAGACCCGCGCCTTCCCGACAATCCAATTATTTTTGCGTCCGATAGCTTCCTGCAATTAACCGAATACAGCCGCGAAGAAATTCTGGGTCGTAATTGTCGCTTCCTTCAGGGGCCAGAGACTGACCGTGCTACGGTACGCAAAATCCGCGACGCAATCGACAATCAAACGGAAGTCACGGTTCAGTTGATTAACTATACGAAGAGCGGAAAAAAATTCTGGAATTTATTTCACTTGCAGCCTATGCGTGACCAGAAGGGCGATGTCCAGTATTTCATTGGCGTTCAGCTTGATGGTACCGAGCATGTTCGCGATGCTGCGGAGCGTGAAGGTGTAATGTTAATTAAAAAGACTGCTGAAAACATTGATGAAGCGGCCAAA' ## since all the sequenced reads have an \"A\" instead of \"G\" at the 413th position, we correct the sequence here\n",
    "\n",
    "LOV = \"LATTLERIEKNFVITDPRLPDNPIIFASDSFLQLTEYSREEILGRNCRFLQGPETDRATVRKIRDAIDNQTEVTVQLINYTKSGKKFWNLFHLQPMRDQKGDVQYFIGVQLDGTEHVRDAAEREGVMLIKKTAENIDEAAK\"\n",
    "\n",
    "gene_len = len(LOV)*3\n",
    "catch_left = ''\n",
    "catch_right = ''\n",
    "\n",
    "\n",
    "\n",
    "Barcodes = {\"BC1_fwd\": \"CCC\", \n",
    "            \"BC1_rev\" : \"CCC\", \n",
    "            \"BC2_fwd\":\t\"AAA\",\n",
    "            \"BC2_rev\":\t\"AAA\"}\n",
    "\n",
    "Primer_seq = {\"_fwd\": \"cgccgCATGGAAGCGAT\".upper(),\n",
    "              \"_rev\":\"cgcacgCGGTTATCCATCG\".upper()}\n",
    "\n",
    "Primer_out_of_triplets = {\"_fwd\": 0, \n",
    "                          \"_rev\": 2}\n",
    "amplicon_seq = \"cgccgcatggaagcgattaacgaaagcagcggtttagccacaacgctggaacgcattgaaaagaatttcgtaatcacagacccgcgccttcccgacaatccaattatttttgcgtccgatagcttcctgcaattaaccgaatacagccgcgaagaaattctgggtcgtaattgtcgcttccttcaggggccagagactgaccgtgctacggtacgcaaaatccgcgacgcaatcgacaatcaaacggaagtcacggttcagttgattaactatacgaagagcggaaaaaaattctggaatttatttcacttgcagcctatgcgtgaccagaagggcgatgtccagtatttcattggcgttcagcttgatggtaccgagcatgttcgcgatgctgcggagcgtgaaggtgtaatgttaattaaaaagactgctgaaaacattgatgaggcggccaaagggagcctgcatccgccgatggataaccgcgtgcg\".upper() #cg\n",
    "\n",
    "genetic_code = {\n",
    "  'ATA': 'I', 'ATC': 'I', 'ATT': 'I', 'ATG': 'M',\n",
    "  'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T',\n",
    "  'AAC': 'N', 'AAT': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "  'AGC': 'S', 'AGT': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "  'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L',\n",
    "  'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P',\n",
    "  'CAC': 'H', 'CAT': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "  'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R',\n",
    "  'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V',\n",
    "  'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A',\n",
    "  'GAC': 'D', 'GAT': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "  'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G',\n",
    "  'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S',\n",
    "  'TTC': 'F', 'TTT': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "  'TAC': 'Y', 'TAT': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "  'TGC': 'C', 'TGT': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "}\n",
    "\n",
    "codons = list(genetic_code.keys())\n",
    "\n",
    "\n",
    "quality_score = {\n",
    "  '!':0, '\"':1, '#':2, '$':3, '%':4, '&':5, \"'\":6, '(':7, ')':8, '*':9,\n",
    "  '+':10, ',':11, '-':12, '.':13, '/':14, '0':15, '1':16, '2':17, '3':18, '4':19,\n",
    "  '5':20, '6':21, '7':22, '8':23, '9':24, ':':25, ';':26, '<':27, '=':28, '>':29,\n",
    "  '?':30, '@':31, 'A':32, 'B':33, 'C':34, 'D':35, 'E':36, 'F':37, 'G':38, 'H':39, 'I':40\n",
    "}\n",
    "\n",
    "\n",
    "ecoli_pref = { ### codons used for retron library (RL8) construction\n",
    "            \"A\": 'GCG',\n",
    "            \"R\": 'CGT',\n",
    "            \"N\": 'AAC',\n",
    "            \"D\": 'GAT',\n",
    "            \"C\": 'TGC',\n",
    "            \"Q\": 'CAG',\n",
    "            \"E\": 'GAA',\n",
    "            \"G\": 'GGC',\n",
    "            \"H\": 'CAT',\n",
    "            \"I\": 'ATT',\n",
    "            \"L\": \"CTG\",\n",
    "            \"K\": 'AAA',\n",
    "            \"M\": 'ATG',\n",
    "            \"F\": \"TTT\",\n",
    "            \"P\": 'CCG',\n",
    "            \"S\": 'AGC',\n",
    "            \"T\": 'ACC',\n",
    "            \"W\": 'TGG',\n",
    "            \"Y\": \"TAT\",\n",
    "            \"V\": 'GTG',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total reads 83748\n",
      "[1, 3, 5, 7, 8, 9, 11, 12, 14, 15]\n",
      "36089\n",
      "2312 b reads are empty\n",
      "538 a reads are empty\n",
      "44891 total reads\n",
      "28166 forward reads with matching BC and primer seq\n",
      "16725 reads with index swapping\n",
      "BC1  28166 reads before filtering\n",
      "################# Done: BC1 \n",
      "################# Done: BC1\n",
      "[0, 2, 4, 6, 10, 13, 16, 17, 18, 19]\n",
      "42635\n",
      "2496 b reads are empty\n",
      "301 a reads are empty\n",
      "52528 total reads\n",
      "35552 forward reads with matching BC and primer seq\n",
      "16976 reads with index swapping\n",
      "BC2  35552 reads before filtering\n",
      "################# Done: BC2 \n",
      "################# Done: BC2\n"
     ]
    }
   ],
   "source": [
    "\n",
    "### make sure to run the second code chunk before this one, since the primer dict could have been updated during analysis of the repeated seq run of S2, S3\n",
    "remove_read_qualities = ['!', '\"', '#', '$', '%', '&', \"'\", '(', ')', '*','+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5']\n",
    "#remove_read_qualities =['!', '\"', '#', '$', '%', '&', \"'\", '(', ')', '*', '+', ',', '-',] \n",
    "#remove_read_qualities =['!', '\"', '#', '$', '%', '&', \"'\", '(', ')', '*', '+', ',', '-', '.', '/']\n",
    "\n",
    "ref_gene = amplicon_seq\n",
    "\n",
    "variant = \"R35\"\n",
    "used_BCs = [\"BC1\", \"BC2\"]\n",
    "\n",
    "## variant 4\n",
    "a_seq, b_seq, _, _, a_ids, b_ids = read_sequences(variant = variant, arbitrary_cutoff_a = False, arbitrary_cutoff_b = False, catch_left=catch_left, catch_right=catch_right, return_qualities_ids=True, quality_score=remove_read_qualities)\n",
    "\n",
    "all_reads, all_ids = demultiplex_reads(a_seq, b_seq, ref_gene = ref_gene ,Barcodes=Barcodes, Primer_seq=Primer_seq, used_Barcodes = used_BCs, Sections = [\"\"], max_mismatch_primerseq = 5, filter_for_n_mut = False, n_mut_treshold = 10, a_ids=a_ids, b_ids=b_ids,  read_len_treshold= None, Primer_out_of_triplets= Primer_out_of_triplets)\n",
    "\n",
    "\n",
    "### names \n",
    "\n",
    "Mut_names = {\"BC1\": \"Mut1\", \n",
    "             \"BC2\" : \"Mut2\"}\n",
    "\n",
    "\n",
    "#all_reads = {**{Mut_names_2[key[:3]]+\"_\"+key[4:] : value for key, value in Cycle2Mutagenesis_reads.items()},\n",
    "#            }\n",
    "\n",
    "# 28156 forward reads with matching BC and primer seq\n",
    "# 16516\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DMS-analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
