{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Deep mutational screen - example notebook \n",
    "We start the analysis from (one or more) amplicon sequencing data sets. These include fastq files for the foward (R1) and respective reverse (R2) reads.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## import packages, the used versions are specified in the requirements.txt file\n",
    "import os\n",
    "from Bio.SeqIO import QualityIO\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.patches as mpatches\n",
    "import gzip\n",
    "import glob\n",
    "import re\n",
    "from DMS_utils import dna_rev_comp, translate_dna2aa\n",
    "import pysam\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import pickle as pkl\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy import stats\n",
    "import os.path\n",
    "from matplotlib.lines import Line2D\n",
    "import json\n",
    "import matplotlib.gridspec as gridspec\n",
    "import shutil\n",
    "from evaluation_functions import *\n",
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### define variables for the analysis\n",
    "base_dir = os.getcwd()\n",
    "\n",
    "ref_seq = \"GGCGCAACTATCGGTATCAAGCTGTTTAAGAAATTCACCTCGAAAGCAAGTTGATAAACTGATACAATTAAAGGCTCCTTTTGGAGCCTTTTTTTTTGGAGTAAGGAGGAAAAATGTCCGAGGAAGCGCAGAACGATCCGCTGCTGCCGGGCTATAGCTTTAACGCGCATCTGGTGGCGGGCCTGACCCCGATTGAAGCGAACGGCTATCTGGATTTTTTTATTGATCGCCCGCTGGGCATGAAAGGCTATATTCTGAACCTGACCATTCGCGGCCAGGGCGTGGTGAAAAACCAGGGCCGCGAATTTGTGTGCCGCCCGGGCGATATTCTGCTGTTTCCGCCGGGCGAAATTCATCATTATGGCCGCCATCCGGAAGCGCGCGAATGGTATCATCAGTGGGTGTATTTTCGCCCGCGCGCGTATTGGCATGAATGGCTGAACTGGCCGAGCATTTTTGCGAACACCGGCTTTTTTCGCCCGGATGAAGCGCATCAGCCGCATTTTAGCGATCTGTTTGGCCAGATTATTAACGCGGGCCAGGGCGAAGGCCGCTATAGCGAACTGCTGGCGATTAACCTGATAGAACAGCTGCTGCTGCGCCGCATGGAAGCGATTAACGAAAGCAGCGGTTTAGCCACAACGCTGGAACGCATTGAAAAGAATTTCGTAATCACAGACCCGCGCCTTCCCGACAATCCAATTATTTTTGCGTCCGATAGCTTCCTGCAATTAACCGAATACAGCCGCGAAGAAATTCTGGGTCGTAATTGTCGCTTCCTTCAGGGGCCAGAGACTGACCGTGCTACGGTACGCAAAATCCGCGACGCAATCGACAATCAAACGGAAGTCACGGTTCAGTTGATTAACTATACGAAGAGCGGAAAAAAATTCTGGAATTTATTTCACTTGCAGCCTATGCGTGACCAGAAGGGCGATGTCCAGTATTTCATTGGCGTTCAGCTTGATGGTACCGAGCATGTTCGCGATGCTGCGGAGCGTGAAGGTGTAATGTTAATTAAAAAGACTGCTGAAAACATTGATGAGGCGGCCAAAGGGAGCCTGCATCCGCCGATGGATAACCGCGTGCGCGAAGCGTGCCAGTATATTAGCGATCATCTGGCGGATAGCAACTTTGATATTGCGAGCGTGGCGCAGCATGTGTGCCTGAGCCCGAGCCGCCTGAGCCATCTGTTTCGCCAGCAGCTGGGCATTAGCGTGCTGAGCTGGCGCGAAGATCAGCGCATTAGCCAGGCGAAACTGCTGCTGAGCACCACCCGCATGCCGATTGCGACCGTGGGCCACAACGTGGGCTTTGATGATCAGCTGTATTTTAGCCGCGTGTTTAAAAAATGCACCGGCGCGAGCCCGAGCGAATTTCGCGCGGGCTGCGAAGAAAAAGTGAACGATGTGGCGGTGAAACTGAGCGGGTAAGGCTAATGGAGATTTTCAACATGGGCTAGCACAGCCCTAGGTATTATGCTAGCGTGGTGTCTGCGTAATAAGGAGTCTTAATCATGCCAGTTC\"\n",
    "\n",
    "catch_left = \"\" #\"ACACTCTTTCCCTACACGACGCTCTTCCGATCT\" ## Adapter fwd\n",
    "catch_right = \"\" #\"GACTGGAGTTCAGACGTGTGCTCTTCCGATCT\"\n",
    "\n",
    "Barcodes = {\n",
    "    \"BC1_fwd\": \"AAGG\",\n",
    "    \"BC1_rev\": \"ATCA\",\n",
    "    \"BC2_fwd\": \"AGTC\",\n",
    "    \"BC2_rev\": \"GCCG\",\n",
    "    \"BC3_fwd\": \"AGGA\",\n",
    "    \"BC3_rev\": \"AATT\",\n",
    "    \"BC4_fwd\": \"GCGA\",\n",
    "    \"BC4_rev\": \"TGGT\",\n",
    "    \"BC5_fwd\": \"CCAT\",\n",
    "    \"BC5_rev\": \"GACC\",\n",
    "    \"BC6_fwd\": \"CTAG\",\n",
    "    \"BC6_rev\": \"CCAG\",\n",
    "    }\n",
    "\n",
    "Primer_seq ={\n",
    "    \"S1_fwd_primer\" : \"GGCGCAACTATCGGTATCAAGC\",\n",
    "    \"S1_rev_primer\" : \"TCGGCCAGTTCAGCCATTCA\",\n",
    "    \"S2_fwd_primer\" : \"ATTATGGCCGCCATCCGGAAGCG\",\n",
    "    \"S2_rev_primer\" : \"TAGCACGGTCAGTCTCTGGC\",\n",
    "    \"S3_fwd_primer\" : \"GTCCGATAGCTTCCTGCAATTAACC\",\n",
    "    \"S3_rev_primer\" : \"ACATGCTGCGCCACGCTC\",\n",
    "    \"S4_fwd_primer\" : \"GATAACCGCGTGCGCGAA\",\n",
    "    \"S4_rev_primer\" : \"GAACTGGCATGATTAAGACTCCTTATTACG\" \n",
    "    }\n",
    "\n",
    "genetic_code = {\n",
    "  'ATA': 'I', 'ATC': 'I', 'ATT': 'I', 'ATG': 'M',\n",
    "  'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T',\n",
    "  'AAC': 'N', 'AAT': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "  'AGC': 'S', 'AGT': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "  'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L',\n",
    "  'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P',\n",
    "  'CAC': 'H', 'CAT': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "  'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R',\n",
    "  'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V',\n",
    "  'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A',\n",
    "  'GAC': 'D', 'GAT': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "  'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G',\n",
    "  'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S',\n",
    "  'TTC': 'F', 'TTT': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "  'TAC': 'Y', 'TAT': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "  'TGC': 'C', 'TGT': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "    }\n",
    "\n",
    "codons = list(genetic_code.keys())\n",
    "\n",
    "quality_score = {\n",
    "  '!':0, '\"':1, '#':2, '$':3, '%':4, '&':5, \"'\":6, '(':7, ')':8, '*':9,\n",
    "  '+':10, ',':11, '-':12, '.':13, '/':14, '0':15, '1':16, '2':17, '3':18, '4':19,\n",
    "  '5':20, '6':21, '7':22, '8':23, '9':24, ':':25, ';':26, '<':27, '=':28, '>':29,\n",
    "  '?':30, '@':31, 'A':32, 'B':33, 'C':34, 'D':35, 'E':36, 'F':37, 'G':38, 'H':39, 'I':40\n",
    "    }\n",
    "\n",
    "Primer_out_of_triplets = {\n",
    "    \"S1_fwd\": 2, ## by how many nucleotides each primer is out of frame\n",
    "    \"S1_rev\": 1, \n",
    "    \"S2_fwd\": 2,\n",
    "    \"S2_rev\": 1,\n",
    "    \"S3_fwd\": 1,\n",
    "    \"S3_rev\": 2,\n",
    "    \"S4_fwd\": 0,\n",
    "    \"S4_rev\": 3\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### read amplicon sequencing reads from fastq files, thereby demultiplexing the reads based on the barcodes and primer sequences \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DMS-analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
