{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio.SeqIO import QualityIO\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "import gzip\n",
    "import glob\n",
    "import re\n",
    "from DMS_utils import dna_rev_comp, translate_dna2aa\n",
    "import pysam\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import pickle as pkl\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy import stats\n",
    "import os.path\n",
    "from matplotlib.lines import Line2D\n",
    "import json\n",
    "import matplotlib.gridspec as gridspec\n",
    "import shutil\n",
    "\n",
    "from evaluation_functions import *\n",
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = os.getcwd()\n",
    "R5_Pool_seq = \"GGCGCAACTATCGGTATCAAGCTGTTTAAGAAATTCACCTCGAAAGCAAGTTGATAAACTGATACAATTAAAGGCTCCTTTTGGAGCCTTTTTTTTTGGAGTAAGGAGGAAAAATGTCCGAGGAAGCGCAGAACGATCCGCTGCTGCCGGGCTATAGCTTTAACGCGCATCTGGTGGCGGGCCTGACCCCGATTGAAGCGAACGGCTATCTGGATTTTTTTATTGATCGCCCGCTGGGCATGAAAGGCTATATTCTGAACCTGACCATTCGCGGCCAGGGCGTGGTGAAAAACCAGGGCCGCGAATTTGTGTGCCGCCCGGGCGATATTCTGCTGTTTCCGCCGGGCGAAATTCATCATTATGGCCGCCATCCGGAAGCGCGCGAATGGTATCATCAGTGGGTGTATTTTCGCCCGCGCGCGTATTGGCATGAATGGCTGAACTGGCCGAGCATTTTTGCGAACACCGGCTTTTTTCGCCCGGATGAAGCGCATCAGCCGCATTTTAGCGATCTGTTTGGCCAGATTATTAACGCGGGCCAGGGCGAAGGCCGCTATAGCGAACTGCTGGCGATTAACCTGATAGAACAGCTGCTGCTGCGCCGCATGGAAGCGATTAACGAAAGCAGCGGTTTAGCCACAACGCTGGAACGCATTGAAAAGAATTTCGTAATCACAGACCCGCGCCTTCCCGACAATCCAATTATTTTTGCGTCCGATAGCTTCCTGCAATTAACCGAATACAGCCGCGAAGAAATTCTGGGTCGTAATTGTCGCTTCCTTCAGGGGCCAGAGACTGACCGTGCTACGGTACGCAAAATCCGCGACGCAATCGACAATCAAACGGAAGTCACGGTTCAGTTGATTAACTATACGAAGAGCGGAAAAAAATTCTGGAATTTATTTCACTTGCAGCCTATGCGTGACCAGAAGGGCGATGTCCAGTATTTCATTGGCGTTCAGCTTGATGGTACCGAGCATGTTCGCGATGCTGCGGAGCGTGAAGGTGTAATGTTAATTAAAAAGACTGCTGAAAACATTGATGAGGCGGCCAAAGGGAGCCTGCATCCGCCGATGGATAACCGCGTGCGCGAAGCGTGCCAGTATATTAGCGATCATCTGGCGGATAGCAACTTTGATATTGCGAGCGTGGCGCAGCATGTGTGCCTGAGCCCGAGCCGCCTGAGCCATCTGTTTCGCCAGCAGCTGGGCATTAGCGTGCTGAGCTGGCGCGAAGATCAGCGCATTAGCCAGGCGAAACTGCTGCTGAGCACCACCCGCATGCCGATTGCGACCGTGGGCCACAACGTGGGCTTTGATGATCAGCTGTATTTTAGCCGCGTGTTTAAAAAATGCACCGGCGCGAGCCCGAGCGAATTTCGCGCGGGCTGCGAAGAAAAAGTGAACGATGTGGCGGTGAAACTGAGCGGGTAAGGCTAATGGAGATTTTCAACATGGGCTAGCACAGCCCTAGGTATTATGCTAGCGTGGTGTCTGCGTAATAAGGAGTCTTAATCATGCCAGTTC\"\n",
    "\n",
    "catch_left = \"ACACTCTTTCCCTACACGACGCTCTTCCGATCT\" ## Adapter fwd\n",
    "catch_right = \"GACTGGAGTTCAGACGTGTGCTCTTCCGATCT\"\n",
    "\n",
    "Barcodes = {\"BC1_Fwd\": \"AAGG\",\n",
    "            \"BC1_Rev\": \"ATCA\",\n",
    "            \"BC2_Fwd\": \"AGTC\",\n",
    "            \"BC2_Rev\": \"GCCG\",\n",
    "            \"BC3_Fwd\": \"AGGA\",\n",
    "            \"BC3_Rev\": \"AATT\",\n",
    "            \"BC4_Fwd\": \"GCGA\",\n",
    "            \"BC4_Rev\": \"TGGT\",\n",
    "            }\n",
    "\n",
    "S1_fwd_primer = \"GGCGCAACTATCGGTATCAAGC\"\n",
    "S1_rev_primer = \"TCGGCCAGTTCAGCCATTCA\"\n",
    "S1_rev_primer_compseq = dna_rev_comp(S1_rev_primer)\n",
    "S2_fwd_primer = \"ATTATGGCCGCCATCCGGAAGCG\"\n",
    "S2_rev_primer = \"TAGCACGGTCAGTCTCTGGC\"\n",
    "S2_rev_primer_compseq = dna_rev_comp(S2_rev_primer)\n",
    "S3_fwd_primer = \"GTCCGATAGCTTCCTGCAATTAACC\"\n",
    "S3_rev_primer = \"ACATGCTGCGCCACGCTC\"\n",
    "S3_rev_primer_compseq = dna_rev_comp(S3_rev_primer)\n",
    "S4_fwd_primer = \"GATAACCGCGTGCGCGAA\"\n",
    "S4_rev_primer = \"GAACTGGCATGATTAAGACTCCTTATTACG\" \n",
    "S4_rev_primer_compseq = dna_rev_comp(S4_rev_primer)\n",
    "\n",
    "Primer_seq = dict(\n",
    "            S1_fwd_primer = \"GGCGCAACTATCGGTATCAAGC\",\n",
    "            S1_rev_primer = \"TCGGCCAGTTCAGCCATTCA\",\n",
    "            S2_fwd_primer = \"ATTATGGCCGCCATCCGGAAGCG\",\n",
    "            S2_rev_primer = \"TAGCACGGTCAGTCTCTGGC\",\n",
    "            S3_fwd_primer = \"GTCCGATAGCTTCCTGCAATTAACC\",\n",
    "            S3_rev_primer = \"ACATGCTGCGCCACGCTC\",\n",
    "            S4_fwd_primer = \"GATAACCGCGTGCGCGAA\",\n",
    "            S4_rev_primer = \"GAACTGGCATGATTAAGACTCCTTATTACG\" \n",
    ")\n",
    "\n",
    "genetic_code = {\n",
    "  'ATA': 'I', 'ATC': 'I', 'ATT': 'I', 'ATG': 'M',\n",
    "  'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T',\n",
    "  'AAC': 'N', 'AAT': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "  'AGC': 'S', 'AGT': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "  'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L',\n",
    "  'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P',\n",
    "  'CAC': 'H', 'CAT': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "  'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R',\n",
    "  'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V',\n",
    "  'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A',\n",
    "  'GAC': 'D', 'GAT': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "  'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G',\n",
    "  'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S',\n",
    "  'TTC': 'F', 'TTT': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "  'TAC': 'Y', 'TAT': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "  'TGC': 'C', 'TGT': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "}\n",
    "\n",
    "codons = list(genetic_code.keys())\n",
    "\n",
    "quality_score = {\n",
    "  '!':0, '\"':1, '#':2, '$':3, '%':4, '&':5, \"'\":6, '(':7, ')':8, '*':9,\n",
    "  '+':10, ',':11, '-':12, '.':13, '/':14, '0':15, '1':16, '2':17, '3':18, '4':19,\n",
    "  '5':20, '6':21, '7':22, '8':23, '9':24, ':':25, ';':26, '<':27, '=':28, '>':29,\n",
    "  '?':30, '@':31, 'A':32, 'B':33, 'C':34, 'D':35, 'E':36, 'F':37, 'G':38, 'H':39, 'I':40\n",
    "}\n",
    "\n",
    "Primer_out_of_triplets = {\"S1_fwd_primer\": 2,\n",
    "                     \"S1_rev_primer\": 1, \n",
    "                     \"S2_fwd_primer\": 2,\n",
    "                     \"S2_rev_primer\": 1,\n",
    "                     \"S3_fwd_primer\": 1,\n",
    "                     \"S3_rev_primer\": 2,\n",
    "                     \"S4_fwd_primer\": 0,\n",
    "                     \"S4_rev_primer\": 3} ##?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_genotype_table_from_AAseqs(AAseqs_dict, ref_AAseq): \n",
    "    \"\"\"\n",
    "    AAseqs_dict = {timepoint: list of AA sequences}\n",
    "    ref_AAseq_dict =reference AA sequence\n",
    "    \"\"\"\n",
    "    genotypes = {}\n",
    "\n",
    "    for timepoint, reads in AAseqs_dict.items():\n",
    "        variants = []\n",
    "        for read in reads: \n",
    "            variants.append(\" \".join([ref_AAseq[i] + str(i) + read[i] for i in range(len(read)) if ref_AAseq[i] != read[i]]))\n",
    "        \n",
    "        #print(variants)\n",
    "        for variant in variants: \n",
    "            if variant == \"\": \n",
    "                variant = \"WT\"\n",
    "            if variant in genotypes: \n",
    "                genotypes[variant][timepoint] += 1\n",
    "            else: \n",
    "                timepoints = {tp : 0 for tp in AAseqs_dict.keys()}\n",
    "                genotypes[variant] = timepoints\n",
    "                genotypes[variant][timepoint] += 1\n",
    "\n",
    "    return genotypes\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total reads 32117\n",
      "BC1\n",
      "total forward reads before filtering 3907\n",
      "total reverse reads before filtering 3907\n",
      "total forward reads after filtering 3079\n",
      "total reverse reads after filtering 3186\n",
      "3907 reads\n",
      "total forward reads before filtering 4347\n",
      "total reverse reads before filtering 4347\n",
      "total forward reads after filtering 3395\n",
      "total reverse reads after filtering 3312\n",
      "4347 reads\n",
      "total forward reads before filtering 5561\n",
      "total reverse reads before filtering 5561\n",
      "total forward reads after filtering 4529\n",
      "total reverse reads after filtering 4236\n",
      "5561 reads\n",
      "total forward reads before filtering 6086\n",
      "total reverse reads before filtering 6086\n",
      "total forward reads after filtering 4956\n",
      "total reverse reads after filtering 5151\n",
      "6086 reads\n",
      "BC2\n",
      "total forward reads before filtering 4193\n",
      "total reverse reads before filtering 4193\n",
      "total forward reads after filtering 3093\n",
      "total reverse reads after filtering 3542\n",
      "4193 reads\n",
      "total forward reads before filtering 3735\n",
      "total reverse reads before filtering 3735\n",
      "total forward reads after filtering 2877\n",
      "total reverse reads after filtering 2746\n",
      "3735 reads\n",
      "total forward reads before filtering 4292\n",
      "total reverse reads before filtering 4292\n",
      "total forward reads after filtering 3373\n",
      "total reverse reads after filtering 3313\n",
      "4292 reads\n",
      "total forward reads before filtering 4172\n",
      "total reverse reads before filtering 4172\n",
      "total forward reads after filtering 3373\n",
      "total reverse reads after filtering 3154\n",
      "4172 reads\n",
      "total reads 20769\n",
      "BC1\n",
      "total forward reads before filtering 1773\n",
      "total reverse reads before filtering 1773\n",
      "total forward reads after filtering 1271\n",
      "total reverse reads after filtering 1377\n",
      "1773 reads\n",
      "total forward reads before filtering 2020\n",
      "total reverse reads before filtering 2020\n",
      "total forward reads after filtering 1483\n",
      "total reverse reads after filtering 1450\n",
      "2020 reads\n",
      "total forward reads before filtering 2656\n",
      "total reverse reads before filtering 2656\n",
      "total forward reads after filtering 1966\n",
      "total reverse reads after filtering 1948\n",
      "2656 reads\n",
      "total forward reads before filtering 2420\n",
      "total reverse reads before filtering 2420\n",
      "total forward reads after filtering 1814\n",
      "total reverse reads after filtering 1922\n",
      "2420 reads\n",
      "BC2\n",
      "total forward reads before filtering 1779\n",
      "total reverse reads before filtering 1779\n",
      "total forward reads after filtering 1266\n",
      "total reverse reads after filtering 1420\n",
      "1779 reads\n",
      "total forward reads before filtering 1804\n",
      "total reverse reads before filtering 1804\n",
      "total forward reads after filtering 1338\n",
      "total reverse reads after filtering 1254\n",
      "1804 reads\n",
      "total forward reads before filtering 2246\n",
      "total reverse reads before filtering 2246\n",
      "total forward reads after filtering 1657\n",
      "total reverse reads after filtering 1717\n",
      "2246 reads\n",
      "total forward reads before filtering 2292\n",
      "total reverse reads before filtering 2292\n",
      "total forward reads after filtering 1704\n",
      "total reverse reads after filtering 1823\n",
      "2292 reads\n",
      "BC3\n",
      "total forward reads before filtering 1181\n",
      "total reverse reads before filtering 1181\n",
      "total forward reads after filtering 778\n",
      "total reverse reads after filtering 953\n",
      "1181 reads\n",
      "total forward reads before filtering 1007\n",
      "total reverse reads before filtering 1007\n",
      "total forward reads after filtering 703\n",
      "total reverse reads after filtering 673\n",
      "1007 reads\n",
      "total forward reads before filtering 1133\n",
      "total reverse reads before filtering 1133\n",
      "total forward reads after filtering 765\n",
      "total reverse reads after filtering 806\n",
      "1133 reads\n",
      "total forward reads before filtering 1084\n",
      "total reverse reads before filtering 1084\n",
      "total forward reads after filtering 804\n",
      "total reverse reads after filtering 816\n",
      "1084 reads\n",
      "BC4\n",
      "total forward reads before filtering 748\n",
      "total reverse reads before filtering 748\n",
      "total forward reads after filtering 513\n",
      "total reverse reads after filtering 581\n",
      "748 reads\n",
      "total forward reads before filtering 621\n",
      "total reverse reads before filtering 621\n",
      "total forward reads after filtering 431\n",
      "total reverse reads after filtering 419\n",
      "621 reads\n",
      "total forward reads before filtering 823\n",
      "total reverse reads before filtering 823\n",
      "total forward reads after filtering 609\n",
      "total reverse reads after filtering 582\n",
      "823 reads\n",
      "total forward reads before filtering 961\n",
      "total reverse reads before filtering 961\n",
      "total forward reads after filtering 743\n",
      "total reverse reads after filtering 709\n",
      "961 reads\n"
     ]
    }
   ],
   "source": [
    "remove_read_qualities = ['!', '\"', '#', '$', '%', '&', \"'\", '(', ')', '*', '+', ',', '-', ]#'.', '/']\n",
    "\n",
    "a_seq, b_seq, _, _, a_ids, b_ids= read_sequences(variant = \"Mutagenesis\", arbitrary_cutoff_a = False, arbitrary_cutoff_b = False, catch_left=\"\", catch_right=\"\", return_qualities_ids=True, quality_score=remove_read_qualities)\n",
    "ref_gene = R5_Pool_seq\n",
    "\n",
    "Mutagenesis_reads, Mut_ids_Dict = demultiplex_reads(a_seq, b_seq, ref_gene,Barcodes=Barcodes, Primer_seq=Primer_seq, used_Barcodes = [\"BC1\", \"BC2\"], Sections = [\"S1\", \"S2\", \"S3\", \"S4\"], max_mismatch_primerseq = 3, filter_for_n_mut = True, n_mut_treshold = 10, a_ids=a_ids, b_ids=b_ids)\n",
    "\n",
    "a_seq, b_seq, _, _, a_ids, b_ids = read_sequences(variant = \"NegPosSelection\", arbitrary_cutoff_a = False, arbitrary_cutoff_b = False, catch_left=\"\", catch_right=\"\", return_qualities_ids=True, quality_score=remove_read_qualities)\n",
    "\n",
    "NegPosSelection_reads, NegPos_ids = demultiplex_reads(a_seq, b_seq, ref_gene,Barcodes=Barcodes, Primer_seq=Primer_seq, used_Barcodes = [\"BC1\", \"BC2\", \"BC3\", \"BC4\"], Sections = [\"S1\", \"S2\", \"S3\", \"S4\"], max_mismatch_primerseq = 3, filter_for_n_mut = True, n_mut_treshold = 10, a_ids=a_ids, b_ids=b_ids)\n",
    "\n",
    "all_reads = {**{\"Mutagenesis_\"+key : value for key, value in Mutagenesis_reads.items()},**{\"NegPosSelection_\"+key : value for key, value in NegPosSelection_reads.items()}}\n",
    "\n",
    "all_ids = {**{\"Mutagenesis_\"+key : value for key, value in Mut_ids_Dict.items()},**{\"NegPosSelection_\"+key : value for key, value in NegPos_ids.items()}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Mutagenesis_BC1_S4_R1', 'Mutagenesis_BC2_S4_R1', 'NegPosSelection_BC1_S4_R1', 'NegPosSelection_BC2_S4_R1', 'NegPosSelection_BC3_S4_R1', 'NegPosSelection_BC4_S4_R1']\n"
     ]
    }
   ],
   "source": [
    "keys = [name for name in all_reads.keys() if \"S4\" in name and \"R1\" in name]\n",
    "print(keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## build genotype table: \n",
    "Section = \"S4\"\n",
    "keys = ['Mutagenesis_BC1_S4_R1', 'NegPosSelection_BC1_S4_R1', 'NegPosSelection_BC2_S4_R1', 'Mutagenesis_BC2_S4_R1', 'NegPosSelection_BC3_S4_R1', 'NegPosSelection_BC4_S4_R1']\n",
    "Barcode_order = [\"BC1\", \"BC1\", \"BC2\", \"BC2\",  \"BC3\",\"BC4\"]\n",
    "AA_seqs_dict = {}\n",
    "\n",
    "for timepoint, step in enumerate(keys):\n",
    "    Barcode = Barcode_order[timepoint]\n",
    "\n",
    "    catch_left = Barcodes[Barcode+\"_Fwd\"]+ Primer_seq[Section+\"_fwd_primer\"][:Primer_out_of_triplets[Section + \"_fwd_primer\"]]\n",
    "    reads = all_reads[step] \n",
    "    reads = [read[read.index(catch_left)+len(catch_left):] for read in reads if catch_left in read]\n",
    "    AAseqs = [translate_dna2aa(read) for read in reads]\n",
    "    AA_seqs_dict[str(timepoint)] =  AAseqs\n",
    "\n",
    "ref_seq_Section = find_reference_seq(ref_gene = R5_Pool_seq, Primer_seq= Primer_seq, Section = \"S4\", Primer_out_of_triplets=Primer_out_of_triplets)\n",
    "ref_AAseq = translate_dna2aa(ref_seq_Section) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes = build_genotype_table_from_AAseqs(AA_seqs_dict, ref_AAseq)\n",
    "genotypes = pd.DataFrame.from_dict(genotypes, orient='index')\n",
    "## calculate as percentage \n",
    "genotypes = genotypes/genotypes.sum()*100\n",
    "\n",
    "## drop rows with less than 1 percent \n",
    "genotypes = genotypes.loc[genotypes.sum(axis=1) > 5]\n",
    "genotypes = genotypes+1e-10\n",
    "\n",
    "adjacency_matrix = pd.DataFrame()\n",
    "adjacency_matrix[\"names\"] = genotypes.index\n",
    "adjacency_matrix[\"parents\"] = [np.nan]+[\"WT\"]*(len(genotypes)-1)\n",
    "\n",
    "\n",
    "adjacency_matrix.to_csv(f\"{base_dir}/adjacency_matrix.csv\", index=False)\n",
    "\n",
    "genotypes.to_csv(f\"{base_dir}/genotype_table.csv\", index = True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DMS-analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
